rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,234 @@
+---
+authors: Anton Miniailo (anton@goteleport.com)
+state: draft
+---
+
+# RFD 135 - Kubernetes App discovery
+
+## Required Approvers
+
+- Engineering: `@r0mant` && `@tigrato`
+- Product: `@klizhentas` || `@xinding33`
+- Security: `@reedloden` || `@jentfoo`
+
+## What
+
+Proposes discovery of apps that are hosted inside Kubernetes clusters and exposed as services.
+
+Related issue: [#25538](https://github.com/gravitational/teleport/issues/25538)
+
+## Why
+
+Some users dynamically deploy apps into their Kubernetes clusters and expose those as Kubernetes services. When they actively
+create/update/remove apps, manually trying to mirror these changes to Teleport may be time-consuming and prone to
+errors, especially for organizations that have a large number of services running in their Kubernetes clusters.
+Teleport will be able to discover Kubernetes service and automatically register them as
+Teleport apps and keep them in up to date state, simplifying process and making user experience much better.
+
+## Scope
+
+In this document we focus on dynamically creating Teleport apps based on discovery of services in Kubernetes clusters that are running
+Teleport agents. Discovery service and app service should be running inside the Kubernetes cluster whose services are being exposed.
+
+## Details
+
+Kubernetes app discovery will build upon existing infrastructure for other discovery capabilities that are already there. We will add another
+type of matchers to the discovery service configuration - `kubernetes`. If these matchers are present, every 5 minutes discovery service will 
+poll Kubernetes cluster, inside which it is running, for a list of services. Based on this list, it will form a list of Teleport apps resources
+and update it on the backend, constantly keeping the dynamic apps state up to date. Writing to the backend will be handled
+by already existing mechanism in our code - reconciler. Reconciler will compare given fresh list of resources and resources we currently have
+on the backend, and then will decide on appropriate actions (create/update/delete) that are required on
+the backend to change resources to the desired state.
+The Teleport app service that runs in the same Kubernetes cluster will then react to those changes, and will start/stop proxying
+apps that are selected by appropriate labels.
+
+Name of the created Teleport app will consist of Kubernetes service name, namespace and Kubernetes cluster name:
+`$SERVICE_NAME-$NAMESPACE-$KUBE_CLUSTER_NAME`. This is done to avoid naming collisions in discovered apps.
```

Comment:
Unfortunately we're pretty restricted in character set - application name should be valid subdomain, so it's alphanumeric + "-". But we can detect name collision among list of discovered apps and exclude both of them with logging an error. Since as you said, it's a corner case with low probability in real usage, I think this should be ok?

Reply:
Given the limited character set I am also unsure how we can prevent collisions.  But it is a corner case, so as long as we are prepared for it in the implementation I have no concerns.  Detection and logging makes sense to me, thank you!